File: 88.js
//LeetCode 88
// Reference: https://leetcode.com/problems/merge-sorted-array/solutions/4722196/simple-beginner-friendly-dry-run-two-pointer-loops-time-o-m-n-space-o-1-gits/?envType=study-plan-v2&envId=top-interview-150

/**
 * @param {number[]} nums1
 * @param {number} m
 * @param {number[]} nums2
 * @param {number} n
 * @return {void} Do not return anything, modify nums1 in-place instead.
 */
var merge = function(nums1, m, nums2, n) {
    if (n === 0) return;

    let len1 = nums1.length;
    let end_idx = len1 - 1;

    while (n > 0 && m > 0){
        if (nums2[n-1] >= nums1[m-1]){

            nums1[end_idx] = nums2[n-1];
            n--;
        } else {
            nums1[end_idx] = nums1[m-1];
            m--;
    }

    end_idx--;
    }

    while (n > 0){
        nums1[end_idx] = nums2[n-1];
        n--;
        end_idx--;
    }
};

File: 88.py
# Leetcode 88 Merged Sorted Array
# Reference: https://www.youtube.com/watch?v=P1Ic85RarKY

class Solution:
    def merge(self, nums1: List[int], m: int, nums2: List[int], n: int) -> None:
        """
        Do not return anything, modify nums1 in-place instead.
        """
        # nums1 last index
        last = m + n - 1

        # Fill in reverse order
        while m > 0 and n > 0: # while there are elements in both nums1 and nums2
            if nums1[m - 1] > nums2[n - 1]:
                nums1[last] = nums1[m - 1]
                m -= 1
            else:
                nums1[last] = nums2[n - 1]
                n -= 1
            last -= 1
        
        # fill nums1 with leftover nums2 elements
        while n > 0:
            nums1[last] = nums2[n - 1]
            n -= 1
            last -= 1